go-controller/pkg/ovn/obj_retry.go

```diff
@@ -1472,51 +1333,49 @@ func (oc *Controller) WatchResource(objectsToRetry *retryObjs) (*factory.Handler
 				}
 				klog.V(5).Infof("Add event received for %s, key=%s", objectsToRetry.oType, key)
 
-				objectsToRetry.initRetryObjWithAdd(obj, key)
-				objectsToRetry.skipRetryObj(key) // prevent iterateRetryResources from processing this entry
-
-				// This only applies to pod watchers (pods + dynamic network policy handlers watching pods):
-				// if ovnkube-master is restarted, it will gets all the add events with completed pods
-				if oc.isObjectInTerminalState(objectsToRetry.oType, obj) {
-					oc.processObjectInTerminalState(objectsToRetry, obj, key, resourceEventAdd)
-					return
-				}
+				objectsToRetry.DoWithLock(key, func(key string) {
+					// This only applies to pod watchers (pods + dynamic network policy handlers watching pods):
+					// if ovnkube-master is restarted, it will get all the add events with completed pods
+					if oc.isObjectInTerminalState(objectsToRetry.oType, obj) {
+						oc.processObjectInTerminalState(objectsToRetry, obj, key, resourceEventAdd)
+						return
+					}
 
-				// If there is a delete entry with the same key, we got an add event for an object
-				// with the same name as a previous object that failed deletion.
-				// Destroy the old object before we add the new one.
-				//
-				// Note it is okay to access retryEntry without lock here, as the entry is either
-				// accessed by iterateRetryResources or in add/update/delete handler of this resource;
-				// the former is prevented as its ignore field is set to true, and the latter is serialized.
-				if retryEntry := objectsToRetry.getObjRetryEntry(key); retryEntry != nil && retryEntry.oldObj != nil {
-					klog.Infof("Detected stale object during new object"+
-						" add of type %s with the same key: %s",
-						objectsToRetry.oType, key)
-					internalCacheEntry := oc.getInternalCacheEntry(objectsToRetry.oType, obj)
-					if err := oc.deleteResource(objectsToRetry, obj, internalCacheEntry); err != nil {
-						klog.Errorf("Failed to delete old object %s of type %s,"+
-							" during add event: %v", key, objectsToRetry.oType, err)
+					retryObj := objectsToRetry.initRetryObjWithAdd(obj, key)
+					// If there is a delete entry with the same key, we got an add event for an object
+					// with the same name as a previous object that failed deletion.
+					// Destroy the old object before we add the new one.
+					//
+					// Note it is okay to access retryEntry without lock here, as the entry is either
+					// accessed by iterateRetryResources or in add/update/delete handler of this resource;
+					// the former is prevented as its ignore field is set to true, and the latter is serialized.
```

Comment:
Is this comment still relevant?

Reply:
dont think so but we can clean it up later